README.md
# bletcd - The Bloglovin' etcd client

Bletcd is a minimalistic etcd client with convenient functionality for watching for changes.

# Usage

```js
var client = bletcd({
  url: 'http://127.0.0.1:4001',
});

var watcher = client.watcher('test/other-key');
watcher.on('error', function(err) {
  console.error('Watch error:', err);
  watcher.stop();
});
watcher.on('change', function(op) {
  console.log('Notified of change:', op);
});

client.put('test/other-key', 'a-value', function(err, response) {
  if (err) {
    console.error('Failed to write value', err);
    watcher.stop();
    return;
  }
  console.log('Wrote a value to etcd');

  client.get('test/other-key', function(err, response) {
    if (err) {
      console.error('Failed to fetch value', err);
    }
    else {
      console.log('Successfully fetched the key value:', response.node.value);
    }

    client.delete('test/other-key', function(err) {
      if (err) {
        console.error('Failed to delete value:', err);
      }
      else {
        console.log('Successfully deleted value');
      }
      watcher.stop();
    });
  });
});
```

Output:
```txt
Notified of change: { action: 'set',
  node:
   { key: '/test/other-key',
     value: 'a-value',
     modifiedIndex: 537,
     createdIndex: 537 } }
Wrote a value to etcd
Successfully fetched the key value: a-value
Notified of change: { action: 'delete',
  node: { key: '/test/other-key', modifiedIndex: 538, createdIndex: 537 },
  prevNode:
   { key: '/test/other-key',
     value: 'a-value',
     modifiedIndex: 537,
     createdIndex: 537 } }
Successfully deleted value
```


CHANGELOG.md
## 2.2.0 (2015-01-14)


#### Bug Fixes

* **main:** guard against unexpected error payload ([304e8b73](https://github.com/bloglovin/bletcd/commit/304e8b732f9bc254ffaccd7e80d85c182db50c62))


#### Features

* **main:** allow separate option for leaderUrl ([541ff6f8](https://github.com/bloglovin/bletcd/commit/541ff6f8577fcd8331e11b47e9e1df299ce2f195))


## 2.1.0 (2015-01-14)


#### Features

* **main:** follow redirects to leader ([79d23ff2](https://github.com/bloglovin/bletcd/commit/79d23ff25ae217227a391449e7ad61cbbcb7944d))


## 2.0.0 (2014-12-09)


#### Features

* **master:** improve the API ([99adfdbe](https://github.com/bloglovin/bletcd/commit/99adfdbe3ca3e4785f770258377d1a9459549e57))


#### Breaking Changes

* This breaks the API for creating a new client, but maintains backwards compatibility for the client object itself.
 ([99adfdbe](https://github.com/bloglovin/bletcd/commit/99adfdbe3ca3e4785f770258377d1a9459549e57))



test/client.js
/* jshint node: true */
/* global suite, test */
'use strict';

var lib = {
  assert: require('assert'),
  nock: require('nock'),
  Etcd: require('../')
};

var etcd = lib.nock('http://127.0.0.1:4001');

etcd.put('/v2/keys/bletcd-test/fooval', 'value=foovalue')
  .reply(200, {'action':'set','node':{'key':'/bletcd-test/fooval','value':'foovalue','modifiedIndex':79154,'createdIndex':79154},'prevNode':{'key':'/bletcd-test/fooval','value':'barvalue','modifiedIndex':79150,'createdIndex':79150}}, { 'content-type': 'application/json',
  'x-etcd-index': '79154',
  'x-raft-index': '921236',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 07:46:42 GMT',
  'transfer-encoding': 'chunked' });

etcd.get('/v2/keys/bletcd-test/fooval')
  .reply(200, {'action':'get','node':{'key':'/bletcd-test/fooval','value':'foovalue','modifiedIndex':79143,'createdIndex':79143}}, { 'content-type': 'application/json',
  'x-etcd-index': '79143',
  'x-raft-index': '920461',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 07:40:20 GMT',
  'transfer-encoding': 'chunked' });

etcd.put('/v2/keys/bletcd-test/fooval', 'value=barvalue')
  .reply(200, {'action':'set','node':{'key':'/bletcd-test/fooval','value':'barvalue','modifiedIndex':79147,'createdIndex':79147},'prevNode':{'key':'/bletcd-test/fooval','value':'foovalue','modifiedIndex':79146,'createdIndex':79146}}, { 'content-type': 'application/json',
  'x-etcd-index': '79147',
  'x-raft-index': '920713',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 07:42:25 GMT',
  'transfer-encoding': 'chunked' });

etcd.get('/v2/keys/bletcd-test/fooval')
  .reply(200, {'action':'get','node':{'key':'/bletcd-test/fooval','value':'barvalue','modifiedIndex':79150,'createdIndex':79150}}, { 'content-type': 'application/json',
  'x-etcd-index': '79150',
  'x-raft-index': '920881',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 07:43:47 GMT',
  'transfer-encoding': 'chunked' });

etcd.delete('/v2/keys/bletcd-test/fooval')
  .reply(200, {'action':'delete','node':{'key':'/bletcd-test/fooval','modifiedIndex':79165,'createdIndex':79164},'prevNode':{'key':'/bletcd-test/fooval','value':'barvalue','modifiedIndex':79164,'createdIndex':79164}}, { 'content-type': 'application/json',
  'x-etcd-index': '79165',
  'x-raft-index': '921790',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 07:51:14 GMT',
  'transfer-encoding': 'chunked' });

etcd.get('/v2/keys/bletcd-test/fooval')
  .reply(404, {'errorCode':100,'message':'Key not found','cause':'/bletcd-test/fooval','index':79187}, { 'content-type': 'application/json',
  'x-etcd-index': '79187',
  date: 'Tue, 29 Jul 2014 07:54:43 GMT',
  'transfer-encoding': 'chunked' });

etcd.post('/v2/keys/bletcd-test/foodir', 'value=nodeval')
  .reply(201, {'action':'create','node':{'key':'/bletcd-test/foodir/79212','value':'nodeval','modifiedIndex':79212,'createdIndex':79212}}, { 'content-type': 'application/json',
  'x-etcd-index': '79212',
  'x-raft-index': '923537',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:05:24 GMT',
  'transfer-encoding': 'chunked' });

etcd.get('/v2/keys/bletcd-test/foodir?recursive=true')
  .reply(200, {'action':'get','node':{'key':'/bletcd-test/foodir','dir':true,'nodes':[{'key':'/bletcd-test/foodir/79212','value':'nodeval','modifiedIndex':79212,'createdIndex':79212}],'modifiedIndex':79212,'createdIndex':79212}}, { 'content-type': 'application/json',
  'x-etcd-index': '79212',
  'x-raft-index': '923537',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:05:24 GMT',
  'transfer-encoding': 'chunked' });

etcd.delete('/v2/keys/bletcd-test/foodir?recursive=true')
  .reply(200, {'action':'delete','node':{'key':'/bletcd-test/foodir','dir':true,'modifiedIndex':79213,'createdIndex':79212},'prevNode':{'key':'/bletcd-test/foodir','dir':true,'modifiedIndex':79212,'createdIndex':79212}}, { 'content-type': 'application/json',
  'x-etcd-index': '79213',
  'x-raft-index': '923538',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:05:24 GMT',
  'transfer-encoding': 'chunked' });

var assert = lib.assert;

suite('CRUD', function testClient() {
  var etcd = lib.Etcd();

  test('Create node', function createNode(done) {
    etcd.set('bletcd-test/fooval', 'foovalue', function createResult(error) {
      done(error);
    });
  });

  test('Get node', function getNode(done) {
    etcd.get('bletcd-test/fooval', function getResult(error, result) {
      if (error) { return done(error); }

      assert.equal(result.node.value, 'foovalue');
      done();
    });
  });

  test('Update node', function createNode(done) {
    etcd.set('bletcd-test/fooval', 'barvalue', function createResult(error) {
      done(error);
    });
  });

  test('Check update of node', function getNode(done) {
    etcd.get('bletcd-test/fooval', function getResult(error, result) {
      if (error) { return done(error); }

      assert.equal(result.node.value, 'barvalue');
      done();
    });
  });

  test('Delete node', function deleteNode(done) {
    etcd.del('bletcd-test/fooval', function delResult(error) {
      done(error);
    });
  });

  test('Check that node is gone', function getNode(done) {
    etcd.get('bletcd-test/fooval', function getResult(error, result) {
      assert.ok(!!error, 'Did not get an error');
      assert.equal(result.errorCode, 100);
      done();
    });
  });

  var sequentialKey;
  test('Create sequentially numbered node', function postNode(done) {
    etcd.create('bletcd-test/foodir', 'nodeval', function createResult(error, result) {
      if (error) { return done(error); }

      sequentialKey = result.node.key;
      done();
    });
  });

  test('Get directory', function getRecursive(done) {
    etcd.get('bletcd-test/foodir', {recursive:true}, function getResult(error, result) {
      if (error) { return done(error); }

      assert.ok(result.node.dir, 'The node is not a directory');
      assert.equal(result.node.nodes.length, 1, 'The node doesn\'t have the expected number of children');

      var child = result.node.nodes[0];
      if (child) {
        assert.equal(child.key, sequentialKey, 'The child has an unknown key');
      }

      done();
    });
  });

  test('Delete directory', function getRecursive(done) {
    etcd.del('bletcd-test/foodir', {recursive:true}, function deleteResult(error) {
      if (error) { return done(error); }
      done();
    });
  });
});


test/watcher.js
/* jshint node: true */
/*global suite, test */
'use strict';

var lib = {
  assert: require('assert'),
  nock: require('nock'),
  Etcd: require('../')
};

var assert = lib.assert;
var etcd = lib.nock('http://127.0.0.1:4001');

etcd.get('/v2/keys/bletcd-test/watched?wait=true').delay(10)
  .reply(200, {'action':'set','node':{'key':'/bletcd-test/watched','value':'foovalue','modifiedIndex':79263,'createdIndex':79263}}, { 'content-type': 'application/json',
  'x-etcd-index': '79262',
  'x-raft-index': '927182',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:35:22 GMT',
  'transfer-encoding': 'chunked' });

etcd.put('/v2/keys/bletcd-test/watched', 'value=foovalue')
  .reply(201, {'action':'set','node':{'key':'/bletcd-test/watched','value':'foovalue','modifiedIndex':79263,'createdIndex':79263}}, { 'content-type': 'application/json',
  'x-etcd-index': '79263',
  'x-raft-index': '927183',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:35:22 GMT',
  'transfer-encoding': 'chunked' });

etcd.delete('/v2/keys/bletcd-test/watched')
  .reply(200, {'action':'delete','node':{'key':'/bletcd-test/watched','modifiedIndex':79264,'createdIndex':79263},'prevNode':{'key':'/bletcd-test/watched','value':'foovalue','modifiedIndex':79263,'createdIndex':79263}}, { 'content-type': 'application/json',
  'x-etcd-index': '79264',
  'x-raft-index': '927184',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:35:22 GMT',
  'transfer-encoding': 'chunked' });

etcd.get('/v2/keys/bletcd-test/watched?wait=true&waitIndex=79264').delay(10)
  .reply(200, {'action':'delete','node':{'key':'/bletcd-test/watched','modifiedIndex':79264,'createdIndex':79263},'prevNode':{'key':'/bletcd-test/watched','value':'foovalue','modifiedIndex':79263,'createdIndex':79263}}, { 'content-type': 'application/json',
  'x-etcd-index': '79264',
  'x-raft-index': '927184',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:35:22 GMT',
  'transfer-encoding': 'chunked' });

suite('Watching', function testClient() {
  var etcd = lib.Etcd();

  test('Watch for changes', function testWatcher(done) {
    var watcher = etcd.watcher('bletcd-test/watched');

    function finished(error) {
      watcher.stop();
      done(error);
    }

    watcher.on('error', finished);

    watcher.on('change', function watcherReacted(op) {
      if (op.action === 'set') {
        etcd.del('bletcd-test/watched', function delResult(error) {
          if (error) { finished(error); }
        });
      }
      else if (op.action === 'delete') {
        finished();
      }
      else {
        finished(new Error('Unexpected change action ' + op.action));
      }
    });

    etcd.set('bletcd-test/watched', 'foovalue', function createResult(error) {
      if (error) { finished(error); }
    });
  });
});

etcd.get('/v2/keys/bletcd-test/watched?wait=true').times(2).reply(500, 'Fuu');

etcd.get('/v2/keys/bletcd-test/watched?wait=true').reply(200, '', {
  'x-etcd-index': '79264'
});

etcd.get('/v2/keys/bletcd-test/watched?wait=true&waitIndex=79265')
  .reply(200, {'action':'set','node':{'key':'/bletcd-test/watched','value':'foovalue','modifiedIndex':79265,'createdIndex':79265}}, { 'content-type': 'application/json',
  'x-etcd-index': '79265',
  'x-raft-index': '927182',
  'x-raft-term': '14',
  date: 'Tue, 29 Jul 2014 08:35:22 GMT',
  'transfer-encoding': 'chunked' });


suite('Watch retry', function testClient() {
  var etcd = lib.Etcd();

  test('Backoff', function testWatcher(done) {
    var watcher = etcd.watcher('bletcd-test/watched');
    var failCount = 0;
    var failT;

    watcher.on('error', function watcherError() {
      if (failCount) {
        watcher.stop();
        var delta = Date.now() - failT;

        if (delta < 300) {
          done(new Error('The watcher is not backing off'));
        }
        else {
          done();
        }
      }
      failCount++;
      failT = Date.now();
    });
  });

  test('Timeout response', function testTimeout(done) {
    var watcher = etcd.watcher('bletcd-test/watched');

    watcher.on('change', function changeAfterRetry(op) {
      assert.equal(op.node.value, 'foovalue');
      watcher.stop();
      done();
    });
  });
});


test/failovers.js
/* jshint node: true */
/* global suite, test */
'use strict';

var lib = {
  nock: require('nock'),
  Etcd: require('../')
};

lib.nock('http://127.0.0.1:4001')
  .put('/v2/keys/bletcd-test/fooval', 'value=foovalue')
  .reply(307, '', {
    'location': 'http://127.0.0.1:4002/v2/keys/bletcd-test/fooval',
    'date': 'Wed, 14 Jan 2015 09:37:39 GMT',
    'content-type': 'text/plain; charset=utf-8',
    'transfer-encoding': 'chunked' });

lib.nock('http://127.0.0.1:4002')
  .put('/v2/keys/bletcd-test/fooval', 'value=foovalue')
  .reply(201, {'action':'set','node':{'key':'/bletcd-test/fooval','value':'foovalue','modifiedIndex':278,'createdIndex':278}}, {
    'content-type': 'application/json',
    'x-etcd-index': '278',
    'x-raft-index': '26391',
    'x-raft-term': '1',
    'date': 'Wed, 14 Jan 2015 09:37:39 GMT',
    'transfer-encoding': 'chunked' });

lib.nock('http://127.0.0.1:4002')
  .put('/v2/keys/bletcd-test/fooval', 'value=foovalue')
  .reply(201, {'action':'set','node':{'key':'/bletcd-test/fooval','value':'foovalue','modifiedIndex':279,'createdIndex':279}}, {
    'content-type': 'application/json',
    'x-etcd-index': '279',
    'x-raft-index': '26392',
    'x-raft-term': '1',
    'date': 'Wed, 14 Jan 2015 09:37:39 GMT',
    'transfer-encoding': 'chunked' });

suite('Failovers', function testClient() {
  var etcd = lib.Etcd();

  test('Client follows redirects to leader', function createNode(done) {
    etcd.set('bletcd-test/fooval', 'foovalue', function createResult(error) {
      done(error);
    });
  });

  test('Next request goes directly to leader', function createNode(done) {
    etcd.set('bletcd-test/fooval', 'foovalue', function createResult(error) {
      done(error);
    });
  });
});


watcher.js
/* jshint node: true */
'use strict';

var events = require('events');

module.exports = createWatcher;

function createWatcher(etcd, key, options) {
  var retryAttempts = 0;
  var requestOpts = {};
  var req;
  var stopped = false;
  var emitter = new events.EventEmitter();

  Object.getOwnPropertyNames(options).forEach(function setOption(name) {
    requestOpts[name] = options[name];
  });

  function wait() {
    if (stopped) { return; }

    requestOpts.wait = true;
    req = etcd.get(key, requestOpts, result);
  }

  function result(error, op, headers) {
    if (stopped) { return; }
    req = undefined;

    if (error) {
      emitter.emit('error', error);

      setTimeout(wait, backOff(retryAttempts));
      retryAttempts++;
    } else if (op === undefined && headers['x-etcd-index']) {
      // Empty timout response from etcd
      requestOpts.waitIndex = parseInt(headers['x-etcd-index'], 10) + 1;
      wait();
    } else {
      retryAttempts = 0;
      requestOpts.waitIndex = op.node.modifiedIndex + 1;

      emitter.emit('change', op);
      wait();
    }
  }

  function backOff(retryAttempts) {
    return Math.pow(2, retryAttempts)*300 + Math.round(Math.random() * 1000);
  }

  function stop() {
    stopped = true;
    if (req) {
      req.abort();
      req = undefined;
    }
  }

  wait();

  return Object.freeze({
    on: function(event, handler) {
      emitter.on(event, handler);
    },
    stop: stop,
  });
}


index.js
/* jshint node: true */
'use strict';

var url = require('url');
var querystring = require('querystring');
var http = require('http');
var https = require('https');

var createWatcher = require('./lib/watcher');
var BLEtcdError = require('./lib/error');

module.exports = createClient;

function createClient(clientOptions) {
  clientOptions = shallowClone(clientOptions) || {};

  var hostUrl = clientOptions.url || 'http://127.0.0.1:4001';
  var host = url.parse(hostUrl);

  var leaderUrl = clientOptions.leaderUrl || hostUrl;
  var leaderHost = url.parse(leaderUrl);

  var apiRoot = '/v2/keys/';

  var agent;
  var ssl = host.protocol === 'https:';
  var httpLib = ssl ? https : http;
  if (ssl || clientOptions.agentConfig) {
    agent = new httpLib.Agent(clientOptions.agentConfig);
  }

  var api = Object.freeze({
    get: get,
    put: put,
    set: put,
    post: post,
    create: post,
    delete: del,
    del: del,
    watcher: watcher,
  });

  function get(key, options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }

    var parameters = shallowClone(options.parameters) || {};
    // Pick up on some options as parameters
    if (options.recursive !== undefined) {
      parameters.recursive = options.recursive;
    }
    if (options.wait !== undefined) {
      parameters.wait = options.wait;
    }
    if (options.waitIndex !== undefined) {
      parameters.waitIndex = options.waitIndex;
    }

    return request('GET', key, {
      parameters: parameters,
      options: options,
    }, callback);
  }

  function put(key, value, options, callback) {
    return write('PUT', key, value, options, callback);
  }

  function post(key, value, options, callback) {
    return write('POST', key, value, options, callback);
  }

  function del(key, options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }

    var parameters = options.parameters || {};
    definedSet(parameters, 'recursive', options.recursive);

    return request('DELETE', key, {
      parameters: parameters,
      options: options,
    }, callback);
  }

  function write(method, key, value, options, callback) {
    if (typeof options === 'function') {
      callback = options;
      options = {};
    }
    var body = shallowClone(options.attributes) || {};
    body.value = value;

    if (options.ttl !== undefined) {
      body.ttl = options.ttl;
    }

    return request(method, key, {
      parameters: shallowClone(options.parameters),
      body: body,
      options: options,
    }, callback);
  }

  function request(method, key, config, callback) {
    var headers = {};
    var options = config.options || {};
    var path = apiRoot + trimKey(key);

    var selectedHost = isWriteOp(method) ? leaderHost : host;

    if (config.parameters !== undefined) {
      var parameters = querystring.stringify(config.parameters);
      if (parameters.length) {
        path += '?' + parameters;
      }
    }

    var payload;
    if (config.body !== undefined) {
      payload = querystring.stringify(config.body);
      headers['Content-type'] = 'application/x-www-form-urlencoded';
      headers['Content-length'] = Buffer.byteLength(payload);
    }

    function performRequest() {
      var req = httpLib.request({
        hostname: selectedHost.hostname,
        port: selectedHost.port || 4001,
        method: method,
        headers: headers,
        path: path,
        agent: agent,
      });

      var timeout = clientOptions.timeout;
      if (options.timeout !== undefined) {
        timeout = options.timeout;
      }
      if (timeout !== undefined) {
        req.setTimeout(timeout);
      }

      if (payload) {
        req.write(payload);
      }

      req.on('response', etcdResponseHandler(function(err, result, headers) {
        if (err && err.reason === BLEtcdError.reason.leaderRedirect) {
          leaderHost = url.parse(headers.location);
          selectedHost = leaderHost;
          currentRequest = performRequest();
          return;
        }
        callback(err, result, headers);
      }));
      req.on('error', function (err) {
        currentRequest = undefined;
        var error = new BLEtcdError('Request failed', BLEtcdError.reason.requestFailed, err);
        callback(error);
      });
      req.end();
    }

    var currentRequest = performRequest();

    return Object.freeze({
      abort: function() {
        if (currentRequest) {
          currentRequest.abort();
        }
      },
    });
  }

  function watcher(key, index, options) {
    if (typeof index === 'object') {
      options = index;
      index = null;
    }

    options = options || {};
    if (typeof index === 'number') {
      options.waitIndex = index;
    }

    return createWatcher(api, key, options);
  }

  return api;
}

function shallowClone(obj) {
  if (!obj || typeof obj !== 'object') {
    return;
  }

  var clone = {};
  Object.getOwnPropertyNames(obj).forEach(function(name) {
    clone[name] = obj[name];
  });
  return clone;
}

function definedSet(target, name, value) {
  if (value !== undefined) {
    target[name] = value;
  }
}

function trimKey(key) {
  return key.replace(/(^\/)|(\/$)/g, '');
}

function isWriteOp(method) {
  return ['POST', 'PUT', 'DELETE'].indexOf(method) !== -1;
}

function etcdResponseHandler(callback) {
  return function onResponse(res) {
    var data = '';

    res.setEncoding('utf8');
    res.on('data', function gotChunk(chunk) {
      data += chunk;
    });

    res.on('end', function gotFullResponse() {
      var payload;
      var error;

      try {
        payload = data.length ? JSON.parse(data) : undefined;
      } catch (err) {
        error = new BLEtcdError('Failed to parse JSON', BLEtcdError.reason.badResponse, err);
        return callback(error, null, res.headers);
      }

      if (res.statusCode === 307) {
        error = new BLEtcdError('Redirected to leader', BLEtcdError.reason.leaderRedirect);
        callback(error, payload, res.headers);
      }
      else if (res.statusCode < 200 || res.statusCode >= 300) {
        var message = payload && payload.message ? payload.message : 'Unknown error';
        error = new BLEtcdError(message, BLEtcdError.reason.errorResponse);
        callback(error, payload, res.headers);
      }
      else {
        callback(null, payload, res.headers);
      }
    });
  };
}
